### PR TITLE
Make our doctools submodule more robust

### PR DESCRIPTION
**Pull request by mpenkov**
- Moved all docstring-related functionality into the doctools submodule
- Make submodules document themselves (use the first line of the docstring to explain what that submodule does, e.g. handle AWS S3)
- Guard against problems in docstring manipulation with general exception (except Exception)
- Add documentation of parse_uri to help

### CODE REVIEW

File: help.txt [hunk 1]
```diff
@@ -219,20 +231,66 @@ FUNCTIONS
...
+        Returns
+        -------
+        collections.namedtuple
+            The parsed URI.
+        
+        Notes
+        -----
+        Supported URI schemes are:
+        
+        * file
```
Comment (piskvorky) [inline help.txt:251]:
```suggestion
        * file (default, used when no scheme specified)
```

File: help.txt [hunk 2]
```diff
@@ -295,12 +353,12 @@ FUNCTIONS
...
 DATA
-    __all__ = ['open', 'smart_open', 's3_iter_bucket', 'register_compresso...
+    __all__ = ['open', 'parse_uri', 'register_compressor', 's3_iter_bucket...
 
 VERSION
     1.10.0
 
 FILE
-    /home/misha/git/smart_open/smart_open/__init__.py
+    /Users/misha/git/smart_open/smart_open/__init__.py
```
Comment (piskvorky) [inline help.txt:362]: Moved from Linux to Mac? :)

Comment (mpenkov) [reply help.txt:362]: From work to home, and back ;)

File: help.txt (hunk 1 above)
Comment (mpenkov) [reply help.txt:251]: I gradually came up with this from scratch myself, but I'm sure it has been done before (e.g. self-documenting plugins).

**Pull request merged**